universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -0,0 +1,143 @@
+package services.program;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import models.Program;
+import repository.ProgramRepository;
+import services.question.QuestionDefinition;
+
+public class ProgramServiceImpl implements ProgramService {
+
+  private ProgramRepository programRepository;
+
+  @Inject
+  public ProgramServiceImpl(ProgramRepository programRepository) {
+    this.programRepository = programRepository;
+  }
+
+  @Override
+  public ImmutableList<ProgramDefinition> listProgramDefinitions() {
+    return programRepository
+        .listPrograms()
+        .toCompletableFuture()
+        .thenApply(
+            programs ->
+                programs.stream()
+                    .map(program -> program.getProgramDefinition())
+                    .collect(ImmutableList.toImmutableList()))
+        .join();
+  }
+
+  @Override
+  public Optional<ProgramDefinition> getProgramDefinition(long id) {
+    return programRepository
+        .lookupProgram(id)
+        .toCompletableFuture()
+        .join()
+        .map(p -> p.getProgramDefinition());
+  }
+
+  @Override
+  public ProgramDefinition createProgramDefinition(String name, String description) {
+    ProgramDefinition programDefinition =
+        ProgramDefinition.builder().setName(name).setDescription(description).build();
+    Program program = new Program(programDefinition);
+    return programRepository.insertProgramSync(program).getProgramDefinition();
+  }
+
+  @Override
+  public ProgramDefinition addBlockToProgram(
+      long programId, String blockName, String blockDescription) throws ProgramNotFoundException {
+    BlockDefinition blockDefinition =
+        BlockDefinition.builder()
+            .setId(1L) // How to autoincrement this block id?
```

Comment:
Oof that's a good question. Since we aren't relying on a database to do this for us, we may need a class for this that can compute the next available ID

Reply:
Or we can generate unique identifiers based on the program ID and the block name or some other identifying information (maybe initial index?)

Reply:
Addressed in chat.